### PR TITLE
Alter `Atomic.modify` to return the result of the supplied action.

### PR DESCRIPTION
**Pull request by andersio**
After revisiting all the use of `Atomic.modify`, it is observed that for those who need the old state, most of them needs only one or two specific bits. However, the current `Atomic.modify` would relentlessly copy everything (or make CoW types non-uniquely referenced).

Therefore, this PR altered `modify` to be more primitive and efficiently covering all the cases. The only downside is the lost of sugar for those who need a full copy of the old value. 🌴 

Having said that, a caveat is that [the Swift compiler can infer the return type of a closure, if the closure has a single expression only](https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20151214/002583.html) at this moment. In other words, the return type of multi-line closures have to be stated explicitly, if it cannot be inferred from the LHS of an assignment statement (or `Void`).

```
// OK, since return type is obviously `Void`.
value.modify { $0 += 1 }
value.modify {
    $0 += 1
    $0 += 2
}
```

```
// This needs explicit type annotation, or the compiler
// will complain about not being able to infer the return
// type.
let previousValue = value.modify {
    let old = $0
    $0 += 1
    return old
}

// This works.
let previousValue = value.modify { value -> Int in
    let old = value
    value += 1
    return old
}

// Or this
let previousValue: Int = value.modify {
    let old = $0
    $0 += 1
    return old
}
```

### CODE REVIEW

File: ReactiveCocoa/Swift/Flatten.swift [hunk 1]
```diff
@@ -429,13 +429,11 @@ private final class ConcatState<Value, Error: Swift.Error> {
...
 
-		var shouldStart = true
-
-		queuedSignalProducers.modify { queue in
+		let shouldStart: Bool = queuedSignalProducers.modify { queue in
 			// An empty queue means the concat is idle, ready & waiting to start
 			// the next producer.
-			shouldStart = queue.isEmpty
-			queue.append(producer)
+			defer { queue.append(producer) }
```
Comment (andersio) [inline ReactiveCocoa/Swift/Flatten.swift]: Shorthand arguments aren't used in any closures with `defer`, which somehow prohibits the use.

File: ReactiveCocoa/Swift/Signal.swift [hunk 2]
```diff
@@ -1286,14 +1296,14 @@ extension SignalProtocol {
...
+				state.modify {
+					guard !$0.values.left.isEmpty && !$0.values.right.isEmpty else {
+						isFinished = $0.isFinished
 						return
 					}
 
-					tuple = (state.values.left.removeFirst(), state.values.right.removeFirst())
-					isFinished = state.isFinished
+					tuple = ($0.values.left.removeFirst(), $0.values.right.removeFirst())
+					isFinished = $0.isFinished
```
Comment (mdiep) [inline ReactiveCocoa/Swift/Signal.swift]: In multi-line cases like this, I think using a named variable _is_ helpful.

**Comment by mdiep** [comment]
This needs a merge as well.

**Comment by andersio** [comment]
The line is set to be linear (usually 1-2 lines anyway) vs divergent control flow.

**Comment by mdiep** [comment]
👍🏻

**Pull request merged**